### DISCUSSION TITLE
How to set custom virtual host in laragon

### DISCUSSION THREAD

**Discussion by rfrkk** [leokhoa/laragon#783] [General]
**For example, I want to create a virtual host and create an httpd-vhosts.conf.**

```
<VirtualHost *:80>
ServerName rabit.localhost
  DocumentRoot "laragon/www/rabit/"
  <Directory "E:/laragon/www/">
    Options +Indexes +Includes +FollowSymLinks +MultiViews
    AllowOverride All
	Require all granted
  </Directory>
</VirtualHost>
```

**I can't find any documentation on how to set it up. Can you teach me how to set it up?**

**Comment by Pen-y-Fan** [comment]
Laragon automatically creates a host entry for the folder name in the www directory. e.g. if you add rabit to E:\laragon\www and restart Laragon, it will detect the folder and create a host entry in C:\Windows\System32\drivers\etc\host:

```ini
127.0.0.1      rabit.test   #laragon magic! 
```

It will automatically create the vhost file too in E:\laragon\etc\apache2\sites-enabled (**Laragon menu > Apache > dir:conf** and select sites-enabled)

You may need to give elevated privileges (permission) to allow Lagaon to add the host entry.

You can access your site using the pretty URL: http://rabit.test

The documentation is under [Pretty URLs](https://laragon.org/docs/pretty-urls.html) in the Laragon docs.

If you prefer to use localhost as your domain change the auto create host name to **{name}.localhost** in Laragon menu > Preferences

Never edit an autogenerated conf file (e.g. auto.rabit.test.conf, if you want a custom conf file, copy the auto version and rename it manual.rabit.test.conf, and adjust as required, you may want to create your own site name or domain and add a host entry too.

**Comment by rfrkk** [comment]
@Pen-y-Fan 
Thank you for your help. I encountered further problems after trying it. Please help me.

### STEP1. copy rabit floder into laragon/www
### STEP2. laragon create a .conf (laragon\etc\apache2\sites-enabled)
![p103](https://github.com/leokhoa/laragon/assets/137764/24803527-9af6-40a3-820a-ab48f6439250)
```
<VirtualHost *:80> 
    DocumentRoot "C:/laragon/www/rabit"
    ServerName rabit.localhost
    ServerAlias *.rabit.localhost
    <Directory "C:/laragon/www/rabit">
        AllowOverride All
        Require all granted
    </Directory>
</VirtualHost>
```
### STEP3. open rabut.localhost, but it go into localhost.

![p104](https://github.com/leokhoa/laragon/assets/137764/d2f665bd-4944-47ba-b159-385be7499a55)

### STEP4. I speculate that Laragon ignores htaccess?
![p105](https://github.com/leokhoa/laragon/assets/137764/b1f6dbf2-ef92-4de3-8f06-2a4abd77f96c)
```
<IfModule mod_rewrite.c>
    <IfModule mod_negotiation.c>
        Options -MultiViews
    </IfModule>
    
    RewriteEngine On
    
    RewriteCond %{REQUEST_FILENAME} -d [OR]
    RewriteCond %{REQUEST_FILENAME} -f
    RewriteRule ^ ^$1 [N]

    RewriteCond %{REQUEST_URI} (\.\w+$) [NC]
    RewriteRule ^(.*)$ public/$1 

    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_FILENAME} !-f
    RewriteRule ^ server.php

</IfModule>

```
```
server.php
<?php

session_start();

/**
 * Laravel - A PHP Framework For Web Artisans
 *
 * @package  Laravel
 * @author   Taylor Otwell <taylor@laravel.com>
 */

$uri = urldecode(
    parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH)
);

// This file allows us to emulate Apache's "mod_rewrite" functionality from the
// built-in PHP web server. This provides a convenient way to test a Laravel
// application without having installed a "real" web server software here.
if ($uri !== '/' && file_exists(__DIR__.'/public'.$uri)) {
    return false;
}

require_once __DIR__.'/public/index.php';
```

**Reply by Pen-y-Fan** [reply]
Check your host file, either click the **h** in the top right Laragon windows or manually open **C:\Windows\System32\drivers\etc\host**

You may need to give Laragon elevated privileges to access and update this system file, look for the flashing shield icon in the navigation window,

**Reply by Pen-y-Fan** [reply]
Apache recognises .htaccess files, once the host file is added.

If you cannot edit the **host** file, you can change the document root, in **preferences** from F:\laragon\www\ to F:\laragon\www\rabit, the access rabit from localhost, without a pretty url.

**Comment by rfrkk** [comment]
@Pen-y-Fan 
Another question about mysql 8.3, please help me.

### I can't change root pw by laragon
![p101](https://github.com/leokhoa/laragon/assets/137764/440849fa-baba-4545-ad5b-df126b137d39)

**Reply by Pen-y-Fan** [reply]
You need to install Laragon 6 for PHP 8 to work properly.

Laragon does not reset the MySQL 8 password automatically, you can go to the command line and manually set one. See [How to reset mysql root password](https://github.com/leokhoa/laragon/discussions/713)

**Comment by rfrkk** [comment]
@Pen-y-Fan 
Thank you very much,it can work fine ! I can throw away the clunky and slow wampserver.
